_posts/2022-06-18-zio-streams.md

```diff
@@ -0,0 +1,537 @@
+use a `Queue`, and with a reference to that `queue` we can offer values to it
+elsewhere. Additional, we can also create a `ZSink` from a `Queue`/`Hub`!
+
+If we can set both the _endpoint_ and the _source_ as a queue, what if we joined
+two `ZStream`s together with the _same_ queue? Let's look at a toy example,
+where we "process" one stream to sanitize values, that we then feed that into
+another stream that now doesn't have to address that concern.
+
+After our `ZStream` practice above, we might first write something like this:
+
+```scala
+  // Bad code!
+  val program: ZIO[Any, Throwable, ExitCode] = for {
+    queue <- Queue.unbounded[Int]
+    producer <- dirtyStream.via(parsePipeline)
+      .run(ZSink.fromQueue(queue))
+    result <- ZStream.fromQueue(queue)
+      .run(ZSink.sum[Int]).debug("sum")
+  } yield ExitCode.success
+```
+
+and it will _almost_ do what we think! It will process the values, and feed them
+to the second `ZStream`, but our program will hang. Why? Because we're working
+with asynchronous things now. Our `result` has processed all the elements from
+our `producer` - but it continues to wait for any new values that could be
+passed into the our `queue`. It can't complete! We need to signal to `result`
+that we should finalize, by closing the `queue`. We could make a `Promise`, and
+use it to `await` before closing the `queue`, but luckily this functionality is
+already included with `ZSink.fromQueueWithShutdown`
+
+Now, we might be tempted to write:
+
+```scala
+  // Still bad code!
+  val program: ZIO[Any, Throwable, ExitCode] = for {
+    queue <- Queue.unbounded[Int]
+    producer <- dirtyStream.via(parsePipeline)
+      .run(ZSink.fromQueueWithShutdown(queue))
+    result <- ZStream.fromQueue(queue)
+      .run(ZSink.sum[Int]).debug("sum")
+  } yield ExitCode.success
+```
+
+and it will work _even less than expected_! We will be greeted by an exception,
+because our queue will be closed before we initialize the `ZStream` of our
+`result` with it! We need to think asynchronously top-to-bottom, and realize
+that we need to `fork` both our `producer` _and_ `result`, and then `join` them,
+so our `queue` is appropriately closed _after_ we've feed all of out data into
+it, and our second stream has opened from it to process the values in it.
+
+```scala
+  val program: ZIO[Any, Throwable, ExitCode] = for {
+    queue <- Queue.unbounded[Int]
+    producer <- dirtyStream.via(parsePipeline)
+      .run(ZSink.fromQueueWithShutdown(queue))
+      .fork
+    result <- ZStream.fromQueue(queue)
+      .run(ZSink.sum[Int]).debug("sum")
+      .fork
+    _ <- producer.join
+    _ <- result.join
+  } yield ExitCode.success
+```
+
+Asynchronous code can be tricky to debug, but becomes a lot easier if you
+recognize the scope of context you are working in - i.e Are we processing the
+content of a file, which is finite, or are we streaming generated events to a
+websocket, which could be infinite.
+
```

Comment:
I really like the progress with this example. It's almost a childish curiosity that I've satisfied, to see what if we plug a source and a sink to the same queue.